File: nRF Toolbox/DFU/FileSelection/FolderFilesViewController.m
/*
 * Copyright (c) 2015, Nordic Semiconductor
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this
 * software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
 * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#import "FolderFilesViewController.h"
#include "UserFilesViewController.h"
#import "AccessFileSystem.h"
#import "DFUViewController.h"

@interface FolderFilesViewController ()

@property (nonatomic, strong)AccessFileSystem *fileSystem;

@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property (weak, nonatomic) IBOutlet UIView *emptyView;

- (IBAction)didClickDone:(id)sender;

@end

@implementation FolderFilesViewController

@synthesize tableView;
@synthesize emptyView;
@synthesize selectedPath;

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.fileSystem = [[AccessFileSystem alloc] init];
    self.navigationItem.title = self.directoryName;
}

-(void)viewWillAppear:(BOOL)animated
{
    self.navigationItem.rightBarButtonItem.enabled = selectedPath != nil;
    [self ensureFolderNotEmpty];
}

- (IBAction)didClickDone:(id)sender {
    NSURL *fileURL = [NSURL fileURLWithPath:selectedPath];
    [self.fileDelegate onFileSelected:fileURL];
    
    // Go back to DFUViewController
    [self dismissViewControllerAnimated:YES completion:^{
        [self.fileDelegate onFileSelected:fileURL];
    }];
    [[UIApplication sharedApplication] setStatusBarStyle:UIStatusBarStyleLightContent];
}

-(void)ensureFolderNotEmpty
{
    if (self.files.count == 0)
    {
        emptyView.hidden = NO;
    }
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.files.count;
}

- (UITableViewCell *)tableView:(UITableView *)tv cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [tv dequeueReusableCellWithIdentifier:@"FolderFilesCell" forIndexPath:indexPath];
    NSString *fileName = [self.files objectAtIndex:indexPath.row];
    NSString *filePath = [self.directoryPath stringByAppendingPathComponent:fileName];
    
    // Configure the cell...
    cell.textLabel.text = [self.files objectAtIndex:indexPath.row];
    if ([self.fileSystem checkFileExtension:fileName fileExtension:@"hex"])
    {
        cell.imageView.image = [UIImage imageNamed:@"ic_file"];
    }
    else if ([self.fileSystem checkFileExtension:fileName fileExtension:@"bin"])
    {
        cell.imageView.image = [UIImage imageNamed:@"ic_file"];
    }
    else if ([self.fileSystem checkFileExtension:fileName fileExtension:@"zip"])
    {
        cell.imageView.image = [UIImage imageNamed:@"ic_archive"];
    }
    
    if ([filePath isEqualToString:selectedPath])
    {
        cell.accessoryType = UITableViewCellAccessoryCheckmark;
    }
    else
    {
        cell.accessoryType = UITableViewCellAccessoryNone;
    }
    
    return cell;
}

-(void)tableView:(UITableView *)tv didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    NSString *fileName = [self.files objectAtIndex:indexPath.row];
    NSString *filePath = [self.directoryPath stringByAppendingPathComponent:fileName];
    NSURL *fileURL = [NSURL fileURLWithPath:filePath];
    
    selectedPath = filePath;
    [tv reloadData];
    self.navigationItem.rightBarButtonItem.enabled = YES;
    
    [self.preselectionDelegate onFilePreselected:fileURL];
}

-(UITableViewCellEditingStyle)tableView:(UITableView *)tableView editingStyleForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return UITableViewCellEditingStyleDelete;
}

-(void)tableView:(UITableView *)tv commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (editingStyle == UITableViewCellEditingStyleDelete)
    {
        NSString *fileName = [self.files objectAtIndex:indexPath.row];
        NSLog(@"Removing file: %@",fileName);
        
        NSString *filePath = [self.directoryPath stringByAppendingPathComponent:fileName];
        [self.fileSystem deleteFile:filePath];
        [self.files removeObjectAtIndex:indexPath.row];
        [tv deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
        
        if ([filePath isEqualToString:selectedPath])
        {
            selectedPath = nil;
            
            [self.preselectionDelegate onFilePreselected:nil];
            self.navigationItem.rightBarButtonItem.enabled = NO;
        }
        
        [self performSelector:@selector(ensureFolderNotEmpty) withObject:nil afterDelay:0.6];
    }
}
@end

File: nRF Toolbox/DFU/FileSelection/FolderFilesViewController.swift
//
//  FolderFilesViewController.swift
//  nRF Toolbox
//
//  Created by Mostafa Berg on 12/05/16.
//  Copyright © 2016 Nordic Semiconductor. All rights reserved.
//

import UIKit

class FolderFilesViewController: UIViewController, UITableViewDataSource, UITableViewDelegate {

    //MARK: - Class Properties
    var files                   : [URL]?
    var directoryPath           : String?
    var directoryName           : String?
    var fileDelegate            : FileSelectionDelegate?
    var preselectionDelegate    : FilePreselectionDelegate?
    var selectedPath            : URL?
    
    //MARK: - View Outlets
    @IBOutlet weak var emptyView: UIView!
    @IBOutlet weak var tableView: UITableView!

    //MARK: - View Actions
    @IBAction func doneButtonTapped(_ sender: AnyObject) {
        doneButtonTappedEventHandler()
    }

    //MARK: - UIViewDelegate
    override func viewDidLoad() {
        super.viewDidLoad()
        if directoryName != nil {
            self.navigationItem.title = directoryName!
        } else {
            self.navigationItem.title = "Files"
        }
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        let enabled = (selectedPath != nil)
        self.navigationItem.rightBarButtonItem?.isEnabled = enabled
        do {
            try self.files = FileManager.default.contentsOfDirectory(at: selectedPath!, includingPropertiesForKeys: nil, options: .skipsHiddenFiles)
        } catch {
            print(error)
        }
        self.ensureDirectoryNotEmpty()
    }

    //MARK: - UITableViewDataSource
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return (files?.count)!
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let aCell = tableView.dequeueReusableCell(withIdentifier: "FolderFilesCell", for:indexPath)
        let aFilePath = files?[indexPath.row]
        let fileName = aFilePath?.lastPathComponent
        
        //Configuring the cell
        aCell.textLabel?.text = fileName
        if fileName?.lowercased().contains(".hex") != false {
            aCell.imageView?.image = UIImage(named: "ic_file")
        } else if fileName?.lowercased().contains(".bin") != false {
            aCell.imageView?.image = UIImage(named: "ic_file")
        } else if fileName?.lowercased().contains(".zip") != false {
            aCell.imageView?.image = UIImage(named: "ic_archive")
        } else {
            aCell.imageView?.image = UIImage(named: "ic_file")
        }
        
        if aFilePath == selectedPath {
            aCell.accessoryType = .checkmark
        } else {
            aCell.accessoryType = .none
        }
        
        return aCell
    }

    //MARK: - UITableViewDelegate
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let filePath = files?[indexPath.row]
        selectedPath = filePath
        tableView.reloadData()
        navigationItem.rightBarButtonItem!.isEnabled = true
        self.preselectionDelegate?.onFilePreselected(withURL: filePath!)
    }
    
    func tableView(_ tableView: UITableView, editingStyleForRowAt indexPath: IndexPath) -> UITableViewCell.EditingStyle {
        return UITableViewCell.EditingStyle.delete
    }
    
    func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCell.EditingStyle, forRowAt indexPath: IndexPath) {
        guard editingStyle == .delete else {
            return
        }

        let filePath = files?[indexPath.row]
        do {
            try FileManager.default.removeItem(at: filePath!)
        } catch {
            print("Error while deleting file: \(error)")
            return
        }

        files?.remove(at: indexPath.row)
        tableView.deleteRows(at: [indexPath], with: .automatic)
    
        if filePath == selectedPath {
            selectedPath = nil
            self.preselectionDelegate?.onFilePreselected(withURL: filePath!)
            self.navigationItem.rightBarButtonItem?.isEnabled = false
        }

        self.ensureDirectoryNotEmpty()

    }
    //MARK: - FolderFilesViewController Implementation
    func ensureDirectoryNotEmpty() {
        if (files?.count)! == 0 {
            emptyView.isHidden = false
        }
    }

    func doneButtonTappedEventHandler(){
        // Go back to DFUViewController
        dismiss(animated: true) { 
            self.fileDelegate?.onFileSelected(withURL: self.selectedPath!)
        }
    }
}
